pkg/informer/informer.go

```diff
@@ -24,92 +24,182 @@ import (
 	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/dynamic/dynamicinformer"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
-
-	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
-	tenancylisters "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
 )
 
 const (
 	resyncPeriod = 10 * time.Hour
-)
 
-type clusterDiscovery interface {
-	WithCluster(name logicalcluster.Name) discovery.DiscoveryInterface
-}
+	groupFirstFoundVersionResourceIndex = "group-firstFoundVersion-resource"
+)
 
 // DynamicDiscoverySharedInformerFactory is a SharedInformerFactory that
 // dynamically discovers new types and begins informing on them.
 type DynamicDiscoverySharedInformerFactory struct {
-	workspaceLister tenancylisters.ClusterWorkspaceLister
-	disco           clusterDiscovery
-	dynamicClient   dynamic.Interface
-	filterFunc      func(interface{}) bool
-	pollInterval    time.Duration
-	indexers        cache.Indexers
+	dynamicClient dynamic.Interface
+	filterFunc    func(interface{}) bool
+	indexers      cache.Indexers
+	crdIndexer    cache.Indexer
 
 	// handlersLock protects multiple writers racing to update handlers.
 	handlersLock sync.Mutex
 	handlers     atomic.Value
 
-	mu               sync.RWMutex
+	updateCh chan struct{}
+
+	informersLock    sync.RWMutex
 	informers        map[schema.GroupVersionResource]informers.GenericInformer
 	startedInformers map[schema.GroupVersionResource]bool
 	informerStops    map[schema.GroupVersionResource]chan struct{}
 	terminating      bool
+	terminatedCh     chan struct{}
+}
+
+// NewDynamicDiscoverySharedInformerFactory returns a factory for shared
+// informers that discovers new types and informs on updates to resources of
+// those types.
+func NewDynamicDiscoverySharedInformerFactory(
+	dynClient dynamic.Interface,
+	filterFunc func(obj interface{}) bool,
+	crdInformer apiextensionsinformers.CustomResourceDefinitionInformer,
+	indexers cache.Indexers,
+) (*DynamicDiscoverySharedInformerFactory, error) {
+	f := &DynamicDiscoverySharedInformerFactory{
+		dynamicClient: dynClient,
+		filterFunc:    filterFunc,
+		indexers:      indexers,
+		crdIndexer:    crdInformer.Informer().GetIndexer(),
+
+		// Use a buffered channel of size 1 to allow enqueuing 1 update notification
+		updateCh: make(chan struct{}, 1),
+
+		informers:        make(map[schema.GroupVersionResource]informers.GenericInformer),
+		startedInformers: make(map[schema.GroupVersionResource]bool),
+		informerStops:    make(map[schema.GroupVersionResource]chan struct{}),
+		terminatedCh:     make(chan struct{}),
+	}
+
+	f.handlers.Store([]GVREventHandler{})
+
+	// Add an index function that indexes a CRD by its group/firstServedVersion/resource. We only need the first
+	// served version because this shared informer factory is expected to be using a wildcard client for partial
+	// metadata only. In this instance, version does not matter, because a wildcard partial metadata list request
```

Comment:
because a wildcard partial metadata list request ... what?

Reply:
I'm tempted to have the constructor take in a rest.Config instead of a client, so we can force the construction the way we need it...

The only way this factory is _currently_ used is with a wildcard partial metadata client. For CRs, when we receive a wildcard partial metadata list, we return all CRs across all clusters, regardless of storage version. See https://github.com/kcp-dev/kubernetes/blob/005ea1354ed5c15eb1f98c1d3c3a7bc6a84c5e17/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/kcp_rest_options_getter.go#L30-L38 for more details.

For the built in types, we aren't currently mixing any storage versions of anything that I'm aware of...

Reply:
I'm basically trying to mimic "ServerPreferredGroupVersions"

Reply:
Thanks for the explanation - mind updating the comment? My first response here was just that you left the paragraph mid-sentence without getting to a point :)

Reply:
heh, got sidetracked. Thanks for noticing!

Reply:
Finished my sentence, PTAL